m2c/evaluate.py

```diff
@@ -860,6 +895,125 @@ def fold(
+    if not (isinstance(right_expr, Literal) and right_expr.value == 31):
+        return expr
+
+    or_left_expr = early_unwrap(left_expr.left)
+    or_right_expr = early_unwrap(left_expr.right)
+    if isinstance(or_left_expr, UnaryOp):
+        arg_expr_left = early_unwrap(or_left_expr.expr)
+        # -x | x
+        if arg_expr_left == or_right_expr:
+            return BinaryOp.icmp(left_expr.right, "!=", Literal(0))
+
+    elif (
+        isinstance(or_left_expr, BinaryOp)
+        and or_left_expr.op == "-"
+        and isinstance(or_right_expr, BinaryOp)
+        and or_right_expr.op == "-"
+    ):
+        # (b - a) | (a - b)
+        if (
+            early_unwrap(or_left_expr.right) == early_unwrap(or_right_expr.left)
+            and early_unwrap(or_left_expr.left) == early_unwrap(or_right_expr.right)
+            and not or_left_expr.is_floating()
+            and not or_right_expr.is_floating()
+        ):
+            return BinaryOp.icmp(or_right_expr.left, "!=", or_right_expr.right)
+
+    return expr
+
+
+def replace_xor_shift(expr: BinaryOp) -> BinaryOp:
+    """
+    Simplify the expressions matching `(((b ^ a) >> 1) - ((b ^ a) & b)) >> 31`
+    into `a < b` or `(((b ^ a) >> 1) - ((b ^ a) & a)) >> 31` into `b < a`,
+    generated code wise the direction of the operator doesn't matter.
+    These are generated by GW MW 2.6 for example.
+    """
+    if expr.is_floating() or expr.op != ">>":
+        return expr
+
+    left_expr = early_unwrap_ints(expr.left)
+    if not (isinstance(left_expr, BinaryOp) and left_expr.op == "-"):
+        return expr
+
+    right_expr = early_unwrap(expr.right)
+    if not (isinstance(right_expr, Literal) and right_expr.value == 31):
+        return expr
+
+    sub_left_expr = early_unwrap(left_expr.left)
+    sub_right_expr = early_unwrap(left_expr.right)
+    if not (
+        isinstance(sub_left_expr, BinaryOp)
+        and early_unwrap(sub_left_expr.right) == Literal(1)
+        and isinstance(sub_right_expr, BinaryOp)
+        and sub_right_expr.op == "&"
+    ):
+        return expr
+
+    sub_left_xor_expr = early_unwrap_ints(sub_left_expr.left)
+    sub_right_xor_expr = early_unwrap_ints(sub_right_expr.left)
+
+    if not (
+        isinstance(sub_left_xor_expr, BinaryOp)
+        and sub_left_xor_expr.op == "^"
+        and sub_left_xor_expr == sub_right_xor_expr
+    ):
+        return expr
+
+    sub_right_right_expr = early_unwrap(sub_right_expr.right)
+
+    if early_unwrap(sub_left_xor_expr.left) == sub_right_right_expr:
+        # (b ^ a) ... & b
+        return BinaryOp.icmp(sub_left_xor_expr.right, "<", sub_left_xor_expr.left)
+    elif early_unwrap(sub_left_xor_expr.right) == sub_right_right_expr:
+        # (b ^ a) ... & a
+        return BinaryOp.icmp(sub_left_xor_expr.left, "<", sub_left_xor_expr.right)
+
+    return expr
+
+
+def replace_shift_add_carry(expr: BinaryOp) -> BinaryOp:
+    """
+    Simplify the expressions matching `((b >> 31) + (a >> 31U)) + M2C_CARRY`
```

Comment:
(Hm, I wish M2C_CARRY expressions kept the addition/subtraction that resulted in the carry, it's tricky to understand why the pattern makes sense right now)